proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,440 @@
+
+    Quantifier  |   Binder
+  --------------+------------
+    forall a.   |
+    Type ->     | x
+
+Zipping stops when binders are exhausted, so the ``forall b.`` does not yield a
+pair. Instead, it becomes a part of the return type. We call the remaining
+quantifiers **trailing**.
+
+During arity inference, it is determined how many of the trailing ``forall``
+quantifiers are included in the arity of a type synonym or type family.
+
+Primary Change: ``@k``-binders
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+1. Relax the syntactic check of ``data``, ``newtype``, ``type``, ``class``,
+   ``type family``, and ``data family`` declarations to allow ``@k``-binders in
+   their headers::
+
+     tv_bndr ::=
+              | v           -- variable
+              | (v :: k)    -- variable with kind annotation
+       (NEW)  | @v          -- invisible variable
+       (NEW)  | @(v :: k)   -- invisible variable with kind annotation
+       (NEW)  | @_          -- skip
+
+2. When a declaration has no standalone kind signature, a ``@k``-binder gives
+   rise to a ``forall k.`` quantifier in the inferred kind signature.
+   The inferred ``forall k.`` does not float to the left; the order of
+   quantifiers continues to match the order of binders in the header.
+
+3. When a standalone kind signature is given, modify the process of zipping
+   binders with quantifiers to match ``@k``-binders and ``forall k.``-quantifiers.
+   To be precise, add the following clause to the pseudo-code of ``zipSAKS``::
+
+     zipSAKS (q:qs) (b:bs)
+       | isInvisibleForall q && isInvisibleBinder b    (NEW)
+             = (q, Just b) : zipSAKS qs bs
+       | ...
+       | ...
+
+
+   ``isInvisibleForall`` holds for the ``forall x.`` quantifier only.
+
+   ``isInvisibleBinder`` holds for ``@k``, ``@(k :: s)``, and ``@_``, but does
+   not hold for ``k`` or ``(k :: s)``.
+
+   This implies that within a single declaration, some ``forall x.``
+   quantifiers may be zipped with binders, while others are not.
+
+Secondary Change: Arity Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+4. During arity inference, do not include trailing ``forall x.`` quantifiers in
+   the arity.
+
+Secondary Change: Scoping Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+5. In type synonym declarations, require that every variable mentioned on the
+   RHS must be bound on the LHS.
+
+6. In type family and data family instances, require that every variable
+   mentioned on the RHS must also occur on the LHS.
+
+Secondary Change: Instantiation Rules
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+7. In type family and data family instances, the instantiation is fully
+   determined by the left hand side, without looking at the right hand side.
+
+Examples
+--------
+
+TBD.
+
+Effect and Interactions
+-----------------------
+
+TBD.
+
+Costs and Drawbacks
+-------------------
+
+TBD.
+
+Alternatives
+------------
+
+* We could also introduce ``@{k}``-binders for ``forall {x}.``-quantifiers, but
```

Comment:
Aha. I'm glad you mention this. So does that mean that it is impossible to bring an inferred variable into scope, other than implicitly via a kind annotation? Perhaps, and perhaps that's OK. Regardless, it might be nice to mention the possibility of inferred variables when talking about zipping.

Reply:
@goldfirere I assume `(..)` from your #238 would apply here too? Then we consistently avoid that problem everywhere.

Reply:
Oh I se you already mentioned above in https://github.com/ghc-proposals/ghc-proposals/pull/425#discussion_r647683861